src/linker/Linker.Steps/BodySubstituterStep.cs

```diff
@@ -0,0 +1,258 @@
+﻿using System;
+using System.IO;
+using System.Xml.XPath;
+using Mono.Cecil;
+using System.Globalization;
+
+namespace Mono.Linker.Steps
+{
+	public class BodySubstituterStep : BaseStep
+	{
+		protected override void Process ()
+		{
+			var files = Context.Substitutions;
+			if (files == null)
+				return;
+
+			foreach (var file in files) {
+				try {
+					ReadSubstitutionFile (GetSubstitutions (file));
+				} catch (Exception ex) when (!(ex is XmlResolutionException)) {
+					throw new XmlResolutionException ($"Failed to process XML substitution '{file}'", ex);
+				}
+
+			}
+		}
+
+		static XPathDocument GetSubstitutions (string substitutionsFile)
+		{
+			using (FileStream fs = File.OpenRead (substitutionsFile)) {
+				return GetSubstitutions (fs);
+			}
+		}
+
+		static XPathDocument GetSubstitutions (Stream substitutions)
+		{
+			using (StreamReader sr = new StreamReader (substitutions)) {
+				return new XPathDocument (new StringReader (sr.ReadToEnd ()));
```

Comment:
Avoid loading the whole stream into a string, then making another reader on top. This should help with performance of the operation and is simpler

```suggestion
				return new XPathDocument (sr);
```

Reply:
Thanks